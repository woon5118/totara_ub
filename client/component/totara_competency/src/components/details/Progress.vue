<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Kevin Hottinger <kevin.hottinger@totaralearning.com>
  @module totara_competency
-->

<template>
  <div class="tui-competencyDetailProgress">
    <ProgressTracker
      :items="formattedForTracker"
      :current-id="myValue"
      :popover-trigger-type="['click']"
      :target-id="minProficientValueId"
    >
      <!-- Popover content -->
      <template v-slot:custom-popover-content="{ description, label, target }">
        <div class="tui-competencyDetailProgress__popover">
          <div
            v-if="target"
            class="tui-competencyDetailProgress__popover-target"
          >
            <Lozenge
              :text="$str('proficient_level', 'totara_competency')"
              type="success"
            />
          </div>

          <h5 class="tui-competencyDetailProgress__popover-header">
            {{ label }}
          </h5>

          <div
            v-if="description"
            class="tui-competencyDetailProgress__popover-body"
            v-html="description"
          />
          <div v-else class="tui-competencyDetailProgress__popover-body">
            {{ $str('no_description', 'totara_competency') }}
          </div>
        </div>
      </template>
    </ProgressTracker>
  </div>
</template>

<script>
// Components
import Lozenge from 'tui/components/lozenge/Lozenge';
import ProgressTracker from 'tui/components/progresstracker/ProgressTracker';

// GraphQL
import ScaleDetailsQuery from 'totara_competency/graphql/scale';

export default {
  components: {
    Lozenge,
    ProgressTracker,
  },
  props: {
    myValue: {
      type: [Number, String],
    },
    competencyId: {
      required: true,
      type: Number,
    },
  },

  data: function() {
    return {
      scale: {},
    };
  },

  apollo: {
    scale: {
      query: ScaleDetailsQuery,
      context: { batch: true },
      variables() {
        return {
          competency_id: this.competencyId,
        };
      },
      update({ totara_competency_scale: scale }) {
        this.$emit('loaded');
        return scale;
      },
    },
  },

  computed: {
    /**
     * Provide ID for the minimum scale at which you can be classed proficient
     *
     * @return {Int}
     */
    minProficientValueId() {
      if (!this.scale.values) return null;
      return this.scale.values.find(({ proficient }) => proficient).id;
    },

    /**
     * Format scale values for progress tracker
     *
     * @return {Array}
     */
    formattedForTracker() {
      if (!this.scale.values) return [];
      return this.scale.values.map(function(elem) {
        return {
          description: elem.description,
          id: elem.id,
          label: elem.name,
        };
      });
    },
  },
};
</script>

<lang-strings>
  {
    "totara_competency": [
      "no_description",
      "proficient_level"
    ]
  }
</lang-strings>

<style lang="scss">
.tui-competencyDetailProgress {
  min-height: 65px;
  margin-top: var(--gap-7);

  &__popover {
    &-header {
      margin-top: 0;
      @include tui-font-heading-label();
    }

    &-body {
      color: var(--color-neutral-6);
    }

    & > * + * {
      margin-top: var(--gap-2);
    }
  }
}
</style>
