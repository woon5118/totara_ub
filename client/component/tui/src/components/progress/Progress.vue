<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Alvin Smith <alvin.smith@totaralearning.com>
  @module tui
-->

<template>
  <div
    ref="progress"
    class="tui-progress"
    :class="{
      'tui-progress--small': small,
      'tui-progress--hideBackground': hideBackground,
    }"
    :aria-valuetext="valueText"
    :aria-valuenow="value"
    :aria-valuemin="min"
    :aria-valuemax="max"
    role="progressbar"
  >
    <div ref="progress-bar" class="tui-progress__bar" :style="progressStyle">
      <span
        ref="progressLabel"
        class="tui-progress__label"
        :class="!hideValue && insideLabel && 'tui-progress__label--inside'"
        >{{ valueText }}</span
      >
    </div>
    <span
      v-show="!hideValue && !insideLabel"
      class="tui-progress__label tui-progress__label--outside"
      >{{ valueText }}</span
    >
  </div>
</template>

<script>
export default {
  props: {
    small: {
      type: Boolean,
      default: false,
    },
    value: {
      type: Number,
      default: 0,
    },
    min: {
      type: Number,
      default: 0,
    },
    max: {
      type: Number,
      default: 100,
    },
    hideValue: {
      type: Boolean,
      default: false,
    },
    format: {
      type: String,
      default: 'percent',
      validator: f => ['number', 'percent'].includes(f),
    },
    completedText: {
      type: [Boolean, String],
      default: false,
    },
    hideBackground: {
      type: Boolean,
      default: false,
    },
    showEmptyState: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      progressStyle: {},
      insideLabel: false,
    };
  },
  computed: {
    fraction() {
      if (this.value > this.max) {
        return 1;
      }
      if (this.value < this.min) {
        return 0;
      }
      return Number((this.value - this.min) / (this.max - this.min));
    },
    valueText() {
      if (this.completedText && this.value === this.max) {
        return typeof this.completedText === 'string'
          ? this.completedText
          : this.$str('completed', 'totara_core');
      }

      if (this.format === 'percent') {
        return `${(this.fraction * 100).toFixed(0)}%`;
      }

      return this.value;
    },
  },
  watch: {
    value: {
      immediate: true,
      handler() {
        this.$_setProgressStyle();

        if (!this.hideValue) {
          this.$nextTick(() => {
            this.$_setLabelStyle();
          });
        }
      },
    },
  },
  methods: {
    $_setLabelStyle() {
      const containerW = this.$refs.progress.offsetWidth;
      const labelW = this.$refs.progressLabel.offsetWidth;

      this.insideLabel = labelW < containerW * this.fraction;
    },
    $_setProgressStyle() {
      let w,
        bgColor = '';

      //handle empty state styling
      if (this.showEmptyState && this.value <= this.min) {
        w = '2%';
        bgColor = '#e6e4e4';
      } else {
        w = `${this.fraction * 100}%`;
      }

      this.progressStyle = {
        width: w,
        backgroundColor: bgColor,
      };
    },
  },
};
</script>

<lang-strings>
  {
    "totara_core": [
      "completed"
    ]
  }
</lang-strings>

<style lang="scss">
.tui-progress {
  @include tui-font-body-xx-small();

  display: flex;
  height: 14px;
  background-color: var(--progress-container-bg-color);
  border: 1px solid var(--progress-border-color);
  border-radius: var(--gap-2);

  &--small {
    height: 12px;
    line-height: 10px;
  }

  &--hideBackground {
    background: none;
  }

  &__bar {
    display: flex;
    justify-content: flex-end;
    background-color: var(--progress-value-bg-color);
    border-radius: var(--gap-2);
    transition: width var(--transition-progress-function)
      var(--transition-progress-duration);
  }

  &__label {
    padding: 0 var(--gap-1);
    opacity: 0;

    &--inside {
      color: var(--progress-value-text-color);
      opacity: 1;
      transition: opacity var(--transition-progress-function)
        var(--transition-progress-duration);
    }
    &--outside {
      color: var(--progress-container-text-color);
      opacity: 1;
    }
  }
}
</style>
