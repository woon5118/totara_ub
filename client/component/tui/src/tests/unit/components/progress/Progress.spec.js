/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Alvin Smith <alvin.smith@totaralearning.com>
 * @module tui
 */

import { mount } from '@vue/test-utils';
import Progress from 'tui/components/progress/Progress';

const factory = propsData => {
  return mount(Progress, {
    propsData: {
      ...propsData,
    },
    mocks: {
      $str: x => x,
    },
  });
};

describe('Progress.vue', () => {
  it('render correctly', () => {
    const wrapper = factory({
      small: false,
      value: 500,
      min: 100,
      max: 1000,
      format: 'percent',
      completedText: true,
      hideBackground: true,
      showEmptyState: true,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('render completed context correctly', () => {
    const wrapper = factory({ value: 100, completedText: true });
    expect(wrapper.find('span').text()).toBe('completed');
  });
});
