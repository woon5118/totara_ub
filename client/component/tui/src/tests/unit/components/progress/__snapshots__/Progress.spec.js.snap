// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress.vue render correctly 1`] = `
<div
  aria-valuemax="1000"
  aria-valuemin="100"
  aria-valuenow="500"
  aria-valuetext="44%"
  class="tui-progress tui-progress--hideBackground"
  role="progressbar"
>
  <div
    class="tui-progress__bar"
    style="width: 44.44444444444444%;"
  >
    <span
      class="tui-progress__label"
    >
      44%
    </span>
  </div>
   
  <span
    class="tui-progress__label tui-progress__label--outside"
  >
    44%
  </span>
</div>
`;
